### ISSUE TITLE
test setResponseCallback doesn't work

### ISSUE THREAD

**Issue by d3QUone** [cocos2d/cocos2d-x#11672]
`HttpResponse setResponseCallback` doesn't work cause the callback function was deprecated. any ideas how to bring it to work now?

**Comment by kdudnyk** [comment]
From the code, you can can use 

``` c
inline void setResponseCallback(const ccHttpRequestCallback& callback)
```

which is not deprecated, for example the usage of this function with lambda would be : 

``` c
network::HttpRequest* request = new (std :: nothrow) network::HttpRequest();
request->setUrl("http://www.google.com");
request->setRequestType(network::HttpRequest::Type::GET);
request->setResponseCallback( [=] (network::HttpClient* client, network::HttpResponse* response) {
    if (!response)
      {
        return;
      }
      // Dump the data
      std::vector<char>* buffer = response->getResponseData();
      for (unsigned int i = 0; i <buffer-> size (); i ++)
      {
        log ("% c", (* buffer) [i]);
      }
});
```

network::HttpClient::getInstance()->sendImmediate(request);
request->release();

Work pretty well for me on v3.6 
What version are you actually using for your tests?

**Comment by d3QUone** [comment]
I use version 3.4. Thank you, your lambda works great! :+1:

**Comment by dariodepaolis** [comment]
And without lambda?

**Comment by kdudnyk** [comment]
I may be wrong, but it's seems that using lambda is how it`s going to be in next versions ...
you can always store your callback as members of ccHttpRequestCallback type in your class

``` c
typedef std::function<void(HttpClient* client, HttpResponse* response)> ccHttpRequestCallback;
```

its actually very simple and works pretty well, you should give it a try

**Comment by dariodepaolis** [comment]
Ok maybe I'm not good, but I don't understand.

I have

```
request->setResponseCallback(CC_CALLBACK_2(MyClass::onPostCompleted, this));
```

And then declared my function onPostCompleted.

Now, do I need to declare onPostCompleted as a ccHttpRequestCallback, or simply create the typedef?

**Comment by kdudnyk** [comment]
ccHttpRequestCallback is a 

```
std::function<void(HttpClient* client, HttpResponse* response)>,
```

( for details about std::function you can reffer to http://www.cplusplus.com/reference/functional/function/ ); 
Shortly - std::function is not a class method, it`s a standalone function

As example you can use it like : 

``` c
class NetworkConnection {
public:
    NetworkConnection() {
        initResponseCallbacks();
    }

    void makeRequest() {
        network::HttpRequest* request = new (std :: nothrow) network::HttpRequest();
        request->setUrl("http://www.google.com");
        request->setRequestType(network::HttpRequest::Type::GET);
        request->setResponseCallback(responseCallback);
        network::HttpClient::getInstance()->sendImmediate(request);
        request->release();
    }

    void initResponseCallbacks() {
        responseCallback1 = [=] (network::HttpClient* client, network::HttpResponse* response) {
            if (!response) {
                return;
            }
            // Dump the data
            std::vector<char>* buffer = response->getResponseData();
            for (unsigned int i = 0; i <buffer-> size (); i ++) {
                log ("% c", (* buffer) [i]);
            }
        };
    }

    ccHttpRequestCallback responseCallback;
};
```

**Comment by walzer** [comment]
Hey guys, please discuss the usage in the forum, not in issue tracker here.
After v3.0, all callbacks are designed as c++11 lambda functions, most of `CCCallback` or similar objects are deprecated.

**Comment by dariodepaolis** [comment]
Ok sorry.
(Thank you kdudnyk)

**Comment by jonnyijapan** [comment]
Sorry to bump again but I get a LOT of warnings when building for Android, because cocos2d-x (3.6) seems to be using the old callback itself... Maybe it's fixed in 3.7 or something, but I'm still on 3.6 and will try to find a way to silence those warnings for whatever in cocos2d-x that is using that. I made sure my game only uses the lambda/block function variant.

**Comment by jonnyijapan** [comment]
Got it.

Instead of just including HttpClient.h, which caused the deprecated warnings, I use this:

```
#pragma GCC diagnostic push
#pragma GCC diagnostic ignored "-Wdeprecated-declarations"
#include "HttpClient.h"
#pragma GCC diagnostic pop
```
